feat: report RTU statistics

higor-melo opened the pull request:
Depends on:
- [ ] https://github.com/tidewise/drivers-modbus/pull/8
- [ ] https://github.com/tidewise/drivers-motors_weg_cvw300/pull/14

# What is this PR for
<!-- A clear description of what this PR do and the changes made.
If you want you can also add the shortcut link here
- [ ] [Shortcut](http://) -->
These report the error counts, total error counts and success count

# How I did it
<!-- Explain a little bit of your implementation -->

# Results, How I tested
<!-- Explain how you tested you PR, if there is a visual output,
     a GIF or image is welcome -->

# Checklist
- [ ] Assign yourself  in the PR
- [ ] Write unit tests (when relevant)
- [ ] Run rubocop and rake tests locally
- [ ] If this PR initialize a CMAKE package please [enable styling and linting](https://github.com/tidewise/wetpaint-buildconf/blob/master/README.md#enabling-styling-and-linting-checks-for-a-cmake-package)
- [ ] Analyse if this PR modifies the UI, if so:
    - [ ] Tested Tupan's simulation (Charts) :world_map:
    - [ ] Tested Tupan's simulation (Hud) :joystick:
    - [ ] Tested ROV's simulation :joystick:

higor-melo committed:
feat: report RTU statistics

These report the error counts, total error counts and
sucess count

higor-melo committed:
fix: write registers return was wrong from the get go

seixasxbr commented:
can't it be tested?

higor-melo commented:
> can't it be tested?


Which part? If you mean the statistics it only forwards the information, however it can be tested, despite the fact that the functionality is already being tested on the base PRs, as for the reply, it's possible to test the method from the helper however I don't see much gain on doing that, thoughts @doudou ?

doudou commented on motors_weg_cvw300.orogen:
```diff
@@ -96,6 +113,8 @@ task_context "Task", subclasses: "iodrivers_base::Task" do
 
     output_port "saturation_signal", "/control_base/SaturationSignal"
 
+    output_port "rtu_statistics", "/modbus/RTUStatistics"
```
documentation

doudou reviewed (commented):
Missing a status period, which has been a common need for these status.

I think the logic in this case should be:
- if the error counters (total error increasing or decreasing) changed since the last status, output unconditionally
- otherwise, output only at the desired period (to have message stats)

higor-melo committed:
feat: allow choosing an update period to avoid flooding logs

This will also always update if an change on any error count
has been detected

higor-melo committed:
feat: add unit tests

fixed typo

higor-melo commented on motors_weg_cvw300.orogen:
```diff
@@ -58,6 +61,26 @@ task_context "Task", subclasses: "iodrivers_base::Task" do
+    # This avoids alternating errors (ERROR, SUCESS, ERROR, SUCESS...) by
+    # by being more sensible to errors since a sucess will always decreaser
+    # the error count by 1 until 0 and it cannot be modified.
+    property "modbus_error_count_increment", "uint8_t", 1
```
There was something about these types I don't quite remember

higor-melo commented on test/task_test.rb:
```diff
@@ -613,6 +613,78 @@
+            expect_execution { have_one_new_sample task.modbus_rtu_statistics_port }
+            tic = Time.now
+
+            assert_in_delta((tic - toc), 2, 0.3)
```
```suggestion
            assert_in_delta((tic - toc), 2, 0.9)
```
Maybe bigger than 1 lower then 3 is a better bet here, specially on CI

higor-melo commented on test/task_test.rb:
```diff
@@ -613,6 +613,78 @@
+        it "sends a rtu statistic if there is an error count change - CRC" do
+            @task.properties.modbus_rtu_statistics_period = Time.at(999)
+            @task.properties.modbus_error_count_threshold = 10
+            # Avoid heisenbug because of the first sample
```
Note: Copy this to the other tests

higor-melo committed:
fix: modbus expect executions need to be handled differently

Read the documentation on modbus_helpers

higor-melo committed:
fix: use the last read message response to test the errors

By doing this we can have a fixed point of synchronization
where we can modify the reply and test if the errors happen
as we expect

higor-melo committed:
feat: validate we catch an extra 0 being sent without throwing

higor-melo commented on test/task_test.rb:
```diff
@@ -613,6 +613,123 @@
+            end.to_have_one_new_sample(task.modbus_rtu_statistics_port)
+
+            assert(task.running?)
+            assert_equal(0, sample.total_crc_error_count)
+            assert_equal(1, sample.total_unexpected_reply_error_count)
```
Missed the `error_count` check (on all of them it seems .-.)
